### PR TITLE
Fixed DiscreteUniform to produce correct density and cdf

### LINKED ISSUE #18611: Incorrect density, cdf calculation in DiscreteUniform
```
>>> from sympy.stats import density, DiscreteUniform, cdf
>>> X = DiscreteUniform('X', [1, 1, 1, 2, 2])
>>> dict(density(X).items())
{1: 1/5, 2: 1/5}
>>> cdf(X)
{1: 1/5, 2: 2/5}
```
Issue is:
Considering that the arguments in the `DiscreteUniformDistribution`, it should be kept as finite set, and therefore should produce `density` as `{1: 1/2, 2: 1/2}`,
Also as `density` is wrongly evaluated, `cdf` produces wrong result, as it should attain value of `1` at some point in the domain of the distribution, which is not satisfied in the above result.

### PR DESCRIPTION
**Pull request by Smit-create**
Fixed DiscreteUniform to produce correct density and cdf
#### References to other Issues or PRs

Fixes #18611

#### Release Notes

* stats
  * `DiscreteUniform` raises ValueError for duplicate args

### CODE REVIEW

**Comment by Smit-create** [comment]
Please review, failing test:
```
________________________________________________________________________________

 sympy/tensor/tests/test_tensor_operators.py:test_expand_partial_derivative_constant_factor_rule 

Traceback (most recent call last):

  File "/home/travis/virtualenv/python3.6.7/lib/python3.6/site-packages/sympy-1.6.dev0-py3.6.egg/sympy/testing/runtests.py", line 1328, in _timeout

    function()

  File "/home/travis/virtualenv/python3.6.7/lib/python3.6/site-packages/sympy-1.6.dev0-py3.6.egg/sympy/tensor/tests/test_tensor_operators.py", line 203, in test_expand_partial_derivative_constant_factor_rule

    nneg*PartialDerivative(A(i), D(j))

AssertionError

 tests finished: 1326 passed, 1 failed, 77 skipped, 185 expected to fail, 

3 expected to fail but passed, in 1258.65 seconds 

DO *NOT* COMMIT!

Traceback (most recent call last):

  File "<stdin>", line 4, in <module>

Exception: Tests failed
```

**Comment by sylee957** [comment]
I guess it's a random test failing.

**Comment by sylee957** [comment]
I wonder what about sanitizing the inputs when creating the classes.

**Comment by Smit-create** [comment]
> sanitizing the inputs

I haven't use it because according to [this comment](https://github.com/sympy/sympy/issues/10104#issuecomment-154070977) DiscreteUnifromDistribution was made to handle general things.

File: sympy/stats/frv_types.py [hunk 1]
```diff
@@ -104,7 +104,7 @@ def set(self):
         return set(self.args)
 
     def pmf(self, x):
-        if x in self.args:
+        if x in set(self.args):
```
Comment (smichr) [inline sympy/stats/frv_types.py]: `set` would not be necessary here.

Comment (Smit-create) [reply sympy/stats/frv_types.py]: Okay, will remove it.

Comment (Smit-create) [reply sympy/stats/frv_types.py]: @smichr Is it good to go now?

**Comment by smichr** [comment]
Should the result for `[1,1,1,2,2]` be `{1:3/5, 2:2/5}`? The documentation is ambiguous: does 'input set' mean 'finite collection' or 'finite set'. If it allows duplicates then that suggests the former. And if that is the case and the distribution is "uniform over the *collection*" then those that appear more often will have a greater probability of appearing.

**Comment by Smit-create** [comment]
I think the DiscreteUniform here was made to work general in the sense to work fine with not only numbers but also symbols. Following the [wikipedia](https://en.wikipedia.org/wiki/Discrete_uniform_distribution), It has an interval as a input, but the current implementation is a bit different from that in wikipedia. Also, IMO input set here tries to indicate the probability distributed unifromly over the unique elements of the input set (finite set), with density as `{1: 1/2, 2: 1/2}`.

**Comment by czgdp1807** [comment]
May be the following can be overridden in `DiscreteUnifromDistribution`.
```python
class SingleFiniteDistribution(Basic, NamedArgsMixin):
    def __new__(cls, *args):
        args = list(map(sympify, args))
        return Basic.__new__(cls, *args)
``` 
The above can reduce the amortized cost of computing results.

**Comment by smichr** [comment]
What type of object would you use to give a distribution that has probabilities `{1:2/3, 2:1/3}`? (I ask because I am not versed in the stats module.)

File: sympy/stats/frv_types.py [hunk 2]
```diff
@@ -92,7 +92,7 @@ def FiniteRV(name, density):
 class DiscreteUniformDistribution(SingleFiniteDistribution):
     @property
     def p(self):
-        return Rational(1, len(self.args))
+        return Rational(1, len(set(self.args)))
```
Comment (smichr) [inline sympy/stats/frv_types.py]:
```suggestion
        return Rational(1, len(self.set))
```

**Comment by sylee957** [comment]
I still think that the issues will ultimately remain if values like `sin(x)**2 + cos(x)**2` and `1` are supplied simultaneously.

**Comment by smichr** [comment]
> issues will ultimately remain

are you commenting on the right issue?

**Comment by sylee957** [comment]
Yes. How would the density like this be.
`Z = DiscreteUniform('Z', [x, 1, sin(y)**2 + cos(y)**2])`
Should it be 1/3 each? Or should it be 1/2?

**Comment by Smit-create** [comment]
> What type of object would you use to give a distribution that has probabilities `{1:2/3, 2:1/3}`? (I ask because I am not versed in the stats module.)
> 
> I would have made this change
> 
> ```diff
> diff --git a/sympy/stats/frv_types.py b/sympy/stats/frv_types.py
> index d5b3560..61e72ec 100644
> --- a/sympy/stats/frv_types.py
> +++ b/sympy/stats/frv_types.py
> @@ -97,7 +97,10 @@ def p(self):
>      @property  # type: ignore
>      @cacheit
>      def dict(self):
> -        return dict((k, self.p) for k in self.set)
> +        d = {k: 0 for k in self.set}
> +        for k in self.args:
> +            d[k] += self.p
> +        return d
> ```
> 
> to give
> 
> ```python
> >>> from sympy.stats import *
> >>> Z = DiscreteUniform('Z', [1, 1, 1, 2, 2, 3, 3, 3, 4])
> >>> dict(density(Z).items())
> {1: 1/3, 2: 2/9, 3: 1/3, 4: 1/9}
> ```

I think this should not work this way because, in `DiscreteUniform` , the probability is not dependent on the frequency of the data, and is distributed uniformly on the input set(finite set).
To obtain density as `{1:2/3, 2:1/3}` you can use:
```
>>> from sympy.stats import *
>>> from sympy import S
>>> X = FiniteRV('X', {1: S(2)/3, 2: S(1)/3})
>>> dict(density(X).items())
{1: 2/3, 2: 1/3}
>>> cdf(X)
{1: 2/3, 2: 1}
>>> Y = Bernoulli('Y', p=S(2)/3, succ=1, fail=2)
>>> dict(density(Y).items())
{1: 2/3, 2: 1/3}
>>> cdf(Y)
{1: 2/3, 2: 1}
```

File: sympy/stats/frv_types.py (hunk 2 above)
Comment (Smit-create) [reply sympy/stats/frv_types.py]: Okay, I will make that change.

**Comment by Smit-create** [comment]
> How would the density like this be.
> `Z = DiscreteUniform('Z', [x, 1, sin(y)**2 + cos(y)**2])`
On the current branch it produces density as `{1: 1/3, x: 1/3, sin(y)**2 + cos(y)**2: 1/3}`
But this can be fixed by overriding `__new__` as suggested by @czgdp1807 .

<pre><b>diff --git a/sympy/stats/frv_types.py b/sympy/stats/frv_types.py</b>
<b>index a0500228d8..3c112e01dd 100644</b>
<b>--- a/sympy/stats/frv_types.py</b>
<b>+++ b/sympy/stats/frv_types.py</b>
<font color="#06989A">@@ -18,8 +18,8 @@</font>
 
 import random
 
<font color="#CC0000">-from sympy import (S, sympify, Rational, binomial, cacheit, Integer,</font>
<font color="#CC0000">-                   Dummy, Eq, Intersection, Interval,</font>
<font color="#4E9A06">+from sympy import (Basic, S, sympify, Rational, binomial, cacheit, Integer,</font>
<font color="#4E9A06">+                   Dummy, Eq, Intersection, Interval, simplify,</font>
                    Symbol, Lambda, Piecewise, Or, Gt, Lt, Ge, Le, Contains)
 from sympy import beta as beta_fn
 from sympy.external import import_module
<font color="#06989A">@@ -90,9 +90,14 @@</font> def FiniteRV(name, density):
     return rv(name, FiniteDistributionHandmade, density)
 
 class DiscreteUniformDistribution(SingleFiniteDistribution):
<font color="#4E9A06">+    def __new__(cls, *args):</font>
<font color="#4E9A06">+        args = list(map(sympify, args))</font>
<font color="#4E9A06">+        args = list(set(map(simplify, args)))</font>
<font color="#4E9A06">+        return Basic.__new__(cls, *args)</font>
<font color="#4E9A06">+</font>
     @property
     def p(self):
<font color="#CC0000">-        return Rational(1, len(set(self.args)))</font>
<font color="#4E9A06">+        return Rational(1, len(self.set))</font>
 
     @property  # type: ignore
     @cacheit
</pre>

From this diff, I get the following results:
```
>>> from sympy.stats import *
>>> from sympy import sin, cos
>>> from sympy.abc import x,y
>>> Z = DiscreteUniform('Z', [x, 1, sin(y)**2 + cos(y)**2])
>>> dict(density(Z).items())
{1: 1/2, x: 1/2}
```
If this changed is good and should I commit it?

**Comment by smichr** [comment]
> If this change is good and should I commit it?

I would definitely advise against this. A distribution should be over things and those things should be math-agnostic.

Seeing that there is a way to give a distribution for a multiset via FiniteRV then I would raise an error in DiscreteUniform if len(set(args)) != len(args) and advise them to use FiniteRV, something like:
```python
ValueError(filldedent('''
Repeated args detected but set expected. If you want a
distribution that has different weightings for each item
consider using DiscreteFinite(%s, %s)''' % symbol, multiset(args))
```

**Comment by Smit-create** [comment]
Sure! Then I would change accordingly.

**Comment by sylee957** [comment]
Okay, I’d agree to this direction, to raise errors for some weird usage

**Comment by smichr** [comment]
Your suggestion in the error string would not have helped them get a weighted distribution.

**Comment by Smit-create** [comment]
Thanks @smichr for adding the error message. The is failing some irrelevant error

**Comment by smichr** [comment]
The error message still looks bad unless the newline is inserted afterwards. Now it looks like this:
```
ValueError:
Repeated args detected but set expected. For a distribution having
different weights for each item use the following:
FiniteRV(X, {a: 1/2, b: 1/3, c: 1/6}).
```
instead of
```
ValueError:
Repeated args detected but set expected. For a distribution having
different weights for each item use the following: FiniteRV(X, {a: 
1/2, b: 1/3, c: 1/6}).
```

**Comment by Smit-create** [comment]
Thanks @smichr

**Pull request merged**